narwhals/backends.py

```diff
@@ -0,0 +1,360 @@
+            elif adapt.version is None:
+                adaptations.extend(replace(adapt, version=v) for v in Version)
+            else:
+                msg = "Adaptation.version must be {Version!r} or None, got {adapt.version!r}"
+                raise TypeError(msg)
+        self.adaptations = adaptations
+
+    def get_adapter(
+        self, cls: type, version: Version = Version.MAIN
+    ) -> Adaptation | None:
+        module_name, *_ = cls.__module__.split(".", maxsplit=1)
+        for adapt in self.adaptations:
+            if adapt.version != version:
+                continue
+
+            if isinstance(adapt.native, type) and cls is adapt.native:
+                return adapt
+
+            elif isinstance(adapt.native, str):
+                adapt.native = cast(str, adapt.native)
+                adapt_module_name, *_, adapt_cls_name = adapt.native.split(".")
+                if (
+                    (adapt_module_name in sys.modules)  # base-module is imported
+                    and (module_name == adapt_module_name)  # roots match
+                    and (cls.__name__ == adapt_cls_name)  # tips match
+                    and (cls is dynamic_import(adapt.native))  # types are identical
+                ):
+                    return adapt
+        return None
+
+    def validate_backend_version(self) -> None:
+        for module_name, version_getter, min_version in self.requires:
+            # TODO(camriddell): this logic may be better suited for a Version namedtuple or dataclass
+            if callable(version_getter):
+                version_str = version_getter()
+            elif isinstance(version_getter, str):
+                version_str = dynamic_import(version_getter)
+            else:
+                msg = "version_getter {version_getter!r} must be a string or callable, got {type(version_getter)}"
+                raise TypeError(msg)
+
+            installed_version = parse_version(version_str)
+            if installed_version < min_version:
+                msg = f"{module_name} must be updated to at least {min_version}, got {installed_version}"
+                raise ValueError(msg)
+
+    def version(self) -> tuple[int, ...]:
+        version_getter = self.requires[0][1]
+        # TODO(camriddell): this logic may be better suited for a Version namedtuple or dataclass
+        if callable(version_getter):
+            version_str = version_getter()
+        elif isinstance(version_getter, str):
+            version_str = dynamic_import(version_getter)
+        else:
+            msg = "version_getter {version_getter!r} must be a string or callable, got {type(version_getter)}"
+            raise TypeError(msg)
+        return parse_version(version_str)
+
+    def native_namespace(self) -> ModuleType:
+        return import_module(self.requires[0][0])
+
+    def get_native_namespace(self) -> ModuleType | None:
+        return sys.modules.get(self.requires[0][0], None)
```

Comment:
General comment, for which I didn't really think of an alternative, is that I don't like too much the fact that `self.requires` is used implicitly 3 times with the first item in the list.

One idea could be to have a distinction between (let's maybe get better names):

- `backend_library` -> this is where the `adaptation.native` class comes from
    - In the proposed implementation this is dynamically extracted from `adaptation.native`, but then for native_namespace the first tuple in `requires` is used
- `extra_requires` -> potential list of extra requirements